### PR TITLE
RF06: Case Sensitivity

### LINKED ISSUE #6030: RF06: Unnecessary Quoted Identifier error when column name needs to be case-sensitive
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When the column name is quoted so that it's case sensitive, the RF06 rule marks the quotes as unnecessary. 

Example Query:
```
SELECT casechange AS "CaseChange"
FROM foo
```

Fix Output:
```
SELECT casechange AS CaseChange
FROM foo
```

### Expected Behaviour

The rule should not fail.

### Observed Behaviour

```
L:   1 | P:  22 | RF06 | Unnecessary quoted identifier "CaseChange".
                       | [references.quoting]
```

### How to reproduce

<details> <summary> test.sql </summary>

```
SELECT casechange AS "CaseChange"
FROM foo
```

</details>

`sqlfluff lint test.sql --dialect ansi`
`sqlfluff fix test_alias.sql --dialect ansi --rules RF06`

### Dialect

ansi

### Version

```
sqlfluff, version 3.1.0
Python 3.11.7
```

### Configuration

NA

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by alanmcruickshank**
This PR resolves #6030, supersedes #6044 and should also fulfil the original intent of #3173 (by effectively reverting #3260).

As per the conversation on slack, there were lots of issues with RF06:
- The docs weren't particularly clear, and made it look like it was only for aliases.
- It wasn't case sensitive (an issue raised by #3173, but which we "fixed" by disabling the rule for snowflake!). The root cause here was a fundamental misunderstanding of the case sensitivity of identifiers in databases.

This PR changes quite a lot:
- The docs are now much more extensive.
- I've removed the `force_enable` config value (and enabled this for snowflake and postres again).
- I've added a now `case_sensitive` config which defaults to `True` (users can override that to `False` to get the old behaviour).
- I've added a bunch more test cases.
- I've also done a small refactor of the way that we can access the `casefold` attribute of parsers, and how we instantiate segments. Should be minor though.

Would appreciate a review from @keraion and @aayushr7 who've been involved with the discussion to get to this level of clarity.

**NOTE:** This PR should trigger a minor release because it changes rule behaviour and config.

### CODE REVIEW

**Comment by alanmcruickshank** [comment]
BTW: Who knew that DuckDB and Spark were both totally case insensitive!?

File: test/fixtures/rules/std_rule_cases/RF06.yml [hunk 1]
```diff
@@ -1,104 +1,106 @@
...
+  # as long as the identifier is *valid* without quotes, we should
+  # remove the quotes.
   fail_str: |
-    SELECT foo
-    FROM [bar].[baz];
+    SELECT
+      "lowercase_reference",
+      "UPPERCASE_REFERENCE",
+      foo AS "column_alias",
+      "Case_Sensitive_Reference",
```
Comment (aayushr7) [inline test/fixtures/rules/std_rule_cases/RF06.yml:62]: I guess this should be `Case_Insensitive_Reference`

Comment (alanmcruickshank) [reply test/fixtures/rules/std_rule_cases/RF06.yml:62]: I see that it's ambiguous 🤔 . My intent here was a reference where the case matters, and therefore can't be unquoted. I'll change it to `Case_Specific_Reference`.

Comment (aayushr7) [reply test/fixtures/rules/std_rule_cases/RF06.yml]: Oh! But the fix is removing the quotes since DuckDB is insensitive, which seems correct.

**Review by aayushr7 — APPROVED** [review]

**Review by keraion — APPROVED** [review]
LGTM. The documentation now definitely clears up the intent of this rule.

**Pull request merged**